GDScript: Fix typed arrays

vonagam opened the pull request:
This is complete rework how typed arrays do function. In terms of interactions now typed arrays act like packed arrays do.

No more implicit conversions between arrays on assigns.
```gdscript
var ints: Array[int] = [1]
var floats: Array[float] = ints # will not work
# same as you cannot do this:
var ints := PackedInt64Array()
var floats: PackedFloat64Array = ints
```

`Array` (without a parameter) for typed arrays is what `Variant` is for other types. It is possible to assign typed array to assignable (variable/parameter/constant) with type of basic array and it is possible to do the opposite, but later fails if types do not match.
```gdscript
var ints: Array[int] = [1]
var basic: Array = ints
ints = basic

var floats: Array[float] = [1.0]
basic = floats
ints = basic # runtime error because assigned is Array[float] and not Array[int]

basic = [2]
ints = basic # runtime error because assigned is Array and not Array[int]
```

Explicit inference no longer tries to guess what element type an array has, it will be typed as basic one. 
```gdscript
var infer := [1, 2, 3] # type is Array, not Array[int]
```

Typed arrays in constants now work properly.
```gdscript
const ints: Array[int] = [1] # value now is really Array[int]
```

There is no implicit conversions between arrays, but there is for array literals. In all examples above literal on right side is a usual array literal, but it gets converted during analysis to be of proper type and fails if it is impossible to do. It applies to variable/parameters/constants initializers, assignments, returns, arguments and casts:
```gdscript
var ints: Array[int] = [1] # init
ints = [2] # assign
var floats := [3.0] as Array[float] # cast
var make_bools = func() -> Array[bool]: return [true] # return
func pass_strings(strings: Array[String]): pass # this is a member function, lambdas are untyped
pass_strings(['ok']) # arg 
```

`typed_assign` now does only one thing - replaces contents of an array with contents from another with conversions if needed. It is also renamed to simple `assign`.
```gdscript
var ints: Array[int] = [1]
var floats: Array[float] = []
floats.assign(ints)
```

Fixed `resize` not using `resize_zeroed` and not initializing new elements in typed arrays. This was leading to memory corruptions and crashes.

In `VariantInterals::initialize` add missing initailization for `Projection` and `Color` (alpha needs to be set at 1).

Added serialization/deserialization of typed arrays. Though it does not handle custom classes at the moment.

Fixed undo/redo for packed arrays, for some reason they were excluded from that before.

Changed editors for arrays/dictionaries a little - no need to do recursively duplicate them on each rerender, do it only when value is about to be modified.

Fixed default values for arrays/dictionaries not showing in editor.

Fixed `GDScriptInstance::set` not converting values when setters involved.

Related issues:
Fixes #53847 (about parameters).
Fixes #54643 (about parameters).
Fixes #55916 (about constants).
Fixes #56389 (about documentation).
Fixes #58285 (about editor).
Fixes #58636 (about parameters).
Fixes #62753 (about editor).
Fixes #63502 (about default values).
Fixes #63570 (about resize).
Fixes #67653 (about editor).
Fixes #67889 (about literals).
Fixes #68978 (about constants).
Fixes #69198 (about constants).
Fixes #69839 (about assigns).
Fixes #70021 (about constants).
Fixes #70235 (about resize).
Fixes #71607 (about editor).
Fixes #71653 (about parameters).
Fixes #71756 (about enums).
Fixes #71815 (about editor).
Fixes #72349 (about parameters).

vonagam committed:
GDScript: Fix issues with typed arrays

vonagam commented:
Rebased and fixed two more problems (one only partially) regarding constant arrays.

Currently:
```gd
func _ready():
  const x: Array[ int ] = [ 1 ]
  var y := x[ 0 ] # error, bad
  x.push_back( '1' )
  print( x ) # [1, "1"], bad
```

First, inference on constant arrays was not working because of misplaced line in subscript reduction that was setting result type to Variant even if type was known (in attribute subscript the line was placed properly, only in index one it was not where it should have been).

Second, partially fixes #68978, the problem was that `reduced_value` of constant arrays was an array on which `set_typed` was not called. Fixed only partially because while now it works for builtin types properly for custom scripts it seems not because `script_type` reference is not resolved yet...

vonagam commented:
Two more commits related to constants. 

First makes all constants (local, member, outer, base) to use `resolve_constant` to resolve the type. Locals were already using it. Members were using copy-paste version of it. Outer/Base ones were only reducing an initializer, but it is insufficient since for constant parameterized arrays a declared type is important. (Also removed useless `all_is_constant` variable from `const_fold_array`/`const_fold_dictionary`.)

Second builds on it and clears up some current problems with deepness/flatness of constants. Now if you index constant by another constant a result is not a constant unless it is happening during constants resolution. Some examples of the problem is shown here - #62680. As I understand though there is a discussion about whenever or not make constants truly deep. Have no problem dropping that commit, but if constants will go as flat into version 4 then this at least makes it work properly.

vonagam commented:
Fixed another thing about arrays (not connected to literals or constants) - typed assigns. Right now if element types are somewhat compatible a destination array just references a source array losing its own type in a process.

```gdscript
class A: pass
class B extends A: pass

func _ready():
  var b: Array[ B ] = [ B.new() ]
  var a: Array[ A ] = b # before: a references b, becomes Array[ B ], after: a copies b contents
  a.push_back( A.new() )
  a.push_back( B.new() )
  print( b.size() ) # before: 2, now: 1
  print( a.size() ) # before: 2, now: 3
  check( a ) # before: error, now: works

func check( a: Array[ A ] ) -> void: print( 'check' )
```

To make it correct array `a` should copy contents of array `b` and maintain its own type info.

By the way, initial commit fixes #69198 and #67889.

Chaosus commented:
You need to squash the commits as required by our pipeline (https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#modifying-a-pull-request).

vonagam commented:
Oh, I was doing that initially, but then it becomes harder to review and also I assume that third commit might be dropped (about flatness/deepness of constants).

vonagam commented:
Fixed #63502, now default values should work properly with params of array type.

Also made array assign be used when typed array is assigned to non-typed one. In `Array::_assign` such case was handled by copying contents, but typed array assign was used only if a destination array had an element type and presence of element type in a source array was ignored (so before typed->untyped was just changing reference, now it copies contents).

Amended last commit, is it ok? Or should I still squash the commits into one?

vonagam commented:
Made changes to `update_array_literal_element_type` and `is_type_compatible`.

In `update_array_literal_element_type` made sure to guard against ending up with Variant in an element type. There was no such guards before but `has_container_element_type` implies that it is not Variant if present.

In `is_type_compatible` reflect that arrays with two different but compatible types are only compatible if implicit conversion is allowed. And same with an assignment of a typed to an untyped array.

Added a test.

vonagam commented:
After I made `is_type_compatible` to use exact equality for element types when no implicit conversion is allowed I've encountered bugs when a class type was not equal to the same class type. Turns out in `is_type_compatible` compassion was done by equating `fqcn`s (fully-qualified class name) while `==` operator only checked equality of pointers to `ClassNode`, but it seems like the same class may have multiple `ClassNode` instances pointing to it (hence, I assume, why `is_type_compatible` used `fqcn`). So I added comparison of `fqcn` in `==` operator for classes types. Right now `fqcn` is a `String`, but is there anything preventing it from being `NodePath`?

vonagam commented:
Fixed another major thing about array assigns.
```gdscript
var first: Array[ A ] = []
var second: Array[ A ] = first
var value: Variant = A.new()
second = [ value ]
print( first.size() ) # before: 1, after: 0
print( second.size() ) # 1
```
Currently `Array::_assign` changes contents of an array it points to if there is a type mismatch instead of creating new array and properly copying contents. Here `second` starts referencing the same array as `first` as they have the same type, but during the assignment of the array literal which is `Array[Variant]` instead of trying to creating new `Array[A]` from what is given it simply replaces an underlying vector of the array it already has.

vonagam commented:
Rebased to resolve conflicts. Fixed one regression from #68747 and another preexisting problem that was revealed there.

The small regression is on [this line](https://github.com/godotengine/godot/blob/f1f8d74b0f0790caab2f8d82ce2130c79b2aaf63/core/variant/array.cpp#L219): when changing the call from `validate` to `validate_object` `Variant::NIL` type was forgotten.

The thing revealed: there were additions of lines in [`GDScriptAnalyzer::is_type_compatible`](https://github.com/godotengine/godot/blob/f1f8d74b0f0790caab2f8d82ce2130c79b2aaf63/modules/gdscript/gdscript_analyzer.cpp#L4186-L4187) and [`Array::_assign`](
https://github.com/godotengine/godot/blob/f1f8d74b0f0790caab2f8d82ce2130c79b2aaf63/core/variant/array.cpp#L206-L207) to make exceptions for `String`/`StringName` pair. In fact those lines are not needed.

In `_assign` it showed that currently there is no check if arrays have types of convertible primitives. Meaning that before you could not assign `Array[float]` to `Array[int]` (where a conversion was allowed). There was no problem converting an array of floats to ints if an array of floats was typed as a simple array of variants, in that case each element would have been checked one by one if they are convertible to int, but no luck if you typed things properly. Now this case is handled.

In `is_type_compatible` [nested call](https://github.com/godotengine/godot/blob/f1f8d74b0f0790caab2f8d82ce2130c79b2aaf63/modules/gdscript/gdscript_analyzer.cpp#L4203) to resolve if element types of arrays are compatible passes false to `p_allow_implicit_conversion` when it should pass the current value down - if an implicit conversion allowed then arrays with convertible types should be ok, if not then not.

vonagam commented:
Squashed into one, removed commit about deepness/flatness of constants, added assertions about previous comment fixes.

Also made code structure of `resolve_parameter` match those of variables and constants (done to properly infer a param type when an array literal is a default value). 

While doing this I noticed that `resolve_variable`, `resolve_constant`, `resolve_parameter` and code for resolving member variables and constants are the samish (it just deciding on a type based on an initializer and an annotated type). It looks like copy-paste with sometimes changed names and missing assertions here and there (I assume because with time people add changes here and there but don't know that the code is duplicated in other places). It looks doable to unify that under a single function (which I did here but only for member constants and local constants). But that just a note, will not do that in this PR.

vonagam commented:
Rebased to fix a conflict with #69992 which was fixing `is_type_compatible` like was written in [the before last comment](https://github.com/godotengine/godot/pull/69248#issuecomment-1345485373). Interestingly it also fixed an inconsistency between member and local constant resolutions but missed the fact that `resolve_parameter` has [the same problem](https://github.com/godotengine/godot/blob/db6976eed4c59b3a320cb26938898c732ef6b7ca/modules/gdscript/gdscript_analyzer.cpp#L1755). This is an example for the last comment. Added missing `true` to `is_type_compatible` check in `resolve_parameter`.

Note on topic of arrays: currently plus operator returns an untyped array even if two arrays involved are typed and have the same type.

kleonc commented on modules/gdscript/tests/scripts/analyzer/features/typed_array_usage.gd:
```diff
@@ -0,0 +1,50 @@
+	var b: Array[B] = [B.new()]
+	print_a(a) # 1
+	# print_b(a)
+	a = b
```
Such assigment shouldn't be allowed. Arrays are reference types so if such assigment would be allowed then it should result in both `a` and `b` referencing the same array (the one `b` is initialized with). But `a` would still be typed as `Array[A]` and `b` would still be typed as `Array[B]`. Meaning the subsequent `a.push_back(A.new())`, although being typewise correct according to the type of `a` and the type of the argument, would be incorrect as the orignal array `a` is referencing is typed as `Array[B]` and shouldn't allow adding elements of a non-`B`-derived types like `A`.

Assigning `Array[Derived]` to a variable (non-`const`) of type `Array[Base]` shouldn't be allowed.

vonagam replied:
Ok, just to be clear. This logic was not introduced by me. It was done properly before in case of assignment of an array of variants to an array of primitives - [here](https://github.com/godotengine/godot/blob/eba33c67b49905973ee7c81a709046415a41e205/core/variant/array.cpp#L227). And it was done improperly in other type mismatches with comments implying that the intention was to do the same - to copy contents.

I was just fixing it, not changing the previously intended behavior.

You saying that it should be changed to forbid such things?

I personally have no real objection to forbidding such assignment, it will make things more strict which generally I do prefer.

anvilfolk replied:
I'd love for this to be clarified also. Are arrays meant to always be reference, or do they follow the "copy on write" policy? If we want them to be reference only, then I agree this needs to be stricter.

Which it wasn't originally! And possibly could just be another PR :)

kleonc replied:
@vonagam To be clear as well: that's just my opinion. I'm not a GDScript maintainer or anything like that. So I'd wait for more opinions on that. But if it was already like that then I think potentially changing it to what I suggested is out of the scope of this PR anyway.

>I was just fixing it, not changing the previously intended behavior.

I'd say the question is whether it was indeed the intended behavior, whether it was thought over at all. And if it's not settled what the behavior should be then I'd say the tests added in this PR shouldn't include such cases. I think tests should be testing what is agreed to be the expected behavior. If there is no agreement on what the expected behavior is in some cases then I see no point in adding tests for such cases (these would just ensure things work like they currently do, which is not necessarily how they're supposed to work).

---

I think that assigment making a copy in case of typed arrays of non matching types is at least surprising. But I think for sure there should be a way to convert (by copying) a typed array of one type to a typed array of another type (like `Array[float]` to `Array[int]` or `Array[B]` -> `Array[A]`). However, currently it seems doing such conversion is possible only by that assignment (CMIIW). So in case it would be decided to be forbidden then some new way to convert typed arrays would need to be added.

---

When an assigment making a copy could make sense? When assigning a const to non-const. Like:
```gdscript
const const_array := [1, 2]
var non_const_array = const_array
```
Of course it could be copy on write, no need to copy right when the assigment is done.

---

All in all currently things are inconsistent, e.g. as mentioned assigning `Array[Derived]` to `Array[Base]` is allowed (and will create a new array) but the same case errors out when it's about passing it to a function:
```gdscript
func foo(a: Array[Base]) -> void:
	pass

func test() -> void:
	var b: Array[Derived] = ...
	foo(b) # Error!
```

---

Sorry for the late response and no real "what to do". To me it feels like we're lacking some kind of general GDScript design document, a guideline every contributor could look into. Seems like we're designing the language in a case by case manner when bug fixing, instead of the language being designed in the first place (then we could be fixing it according to the design). But I think it's a little too late to do anything about it now. Oh, well. :slightly_smiling_face:

Back to your PR: I haven't really looked into the code, just checked the tests and said about what catched my eye. Regarding the GDScript code I think e.g. @anvilfolk, @adamscott, or @rune-scape are more into it / are more familiar with it than I am, and hence probably will be better reviewers.
So yeah, feel free to don't do anything in this PR about the issue I've mentioned, it can be always tackled in another PR.

vonagam replied:
Completely agree with what you said. I just did not question that `_assign` does two (currently three, but one is obviously wrong) different things - changes reference or changes contents.

I think `typed_assign` method should be about changing contents and `=` operator on a variable should be about changing reference.

vonagam replied:
With new behavior it is no longer allowed. Resolved.

anvilfolk commented on core/variant/array.cpp:
```diff
@@ -202,53 +202,77 @@ uint32_t Array::recursive_hash(int recursion_count) const {
 bool Array::_assign(const Array &p_array) {
-	bool can_convert = p_array._p->typed.type == Variant::NIL;
-	can_convert |= _p->typed.type == Variant::STRING && p_array._p->typed.type == Variant::STRING_NAME;
-	can_convert |= _p->typed.type == Variant::STRING_NAME && p_array._p->typed.type == Variant::STRING;
```
I would be careful removing these lines! :) This was recently added by @rune-scape to ensure that `String` and `StringName` are unified. Maybe they'll have some more input on this.

vonagam replied:
Here is [a comment with explanation](https://github.com/godotengine/godot/pull/69248#issuecomment-1345485373).

anvilfolk replied:
I see - so because `can_convert_strict` works between `String` and `StringNames`, it isn't an issue? :)

vonagam replied:
Correct.

anvilfolk commented on core/variant/array.cpp:
```diff
@@ -202,53 +202,77 @@ uint32_t Array::recursive_hash(int recursion_count) const {
+	if (typed.type == Variant::NIL) {
+		// from anything to variants
+		for (int i = 0; i < size; i++) {
+			data[i] = source[i];
```
Is there no other way to get a copy of an array? And if you have to iterate over the values, why not use the `Vector` directly rather than using the pointers?

vonagam replied:
It is the fastest direct option and `Vector` methods just provide safe-guard checks that index is less than a size. Since we know `size` there is no risk there.

anvilfolk replied:
I see, so this will be a little more efficient than using the array methods directly. Might be worth putting that in a comment? :)

anvilfolk commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1734,58 +1687,61 @@ void GDScriptAnalyzer::resolve_match_pattern(GDScriptParser::PatternNode *p_matc
 void GDScriptAnalyzer::resolve_parameter(GDScriptParser::ParameterNode *p_parameter) {
-	GDScriptParser::DataType result;
-	result.kind = GDScriptParser::DataType::VARIANT;
+	GDScriptParser::DataType type;
```
I actually like that this is named result, and I think that's the pattern elsewhere as well, so I'd personally avoid renaming here.

vonagam replied:
It is called `type` in `resolve_variable` and `resolve_constant`. The other places that this code has to be consistent with and potentially merged with in the future.

anvilfolk replied:
My bad then :) I think a lot of the `reduce_*` functions I was working with used `result`, that's why I made this comment! But if the `resolve_*` functions are now more standardized, perfect!

anvilfolk commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1929,49 +1885,50 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
+	bool has_base_element_type = p_base_type != nullptr && p_base_type->has_container_element_type();
+	if (has_base_element_type) {
+		base_element_type = p_base_type->get_container_element_type();
+		base_element_type.is_constant = false;
```
Would it be appropriate to use `type_from_metatype` here?

vonagam replied:
Have not encountered this method call in any places of type resolution. Seems like it does not do much except removing info about original enum. You saying that `[Enum.A, Enum.B]` should be inferred as `Array[int]` and not `Array[Enum]`?

anvilfolk replied:
I'm not really referring to enums - just in general to "metatypes".

It's a weird method. I tend to think of a `metatype` as the `DataType` referring to the class itself, whereas if it isn't a metatype, then it refers to an instance of that class. Similarly, are you referring to the `enum` itself, or to a variable with a type of that enum.

So, if the base element type is `class A`, which is a `metatype` and incidentally also `constant`, then you can use `type_from_metatype` to say that elements are instances of the class `A`. Otherwise you're saying elements of the array are the class itself, not instances of the class.

Though to be fair, it almost seems like `is_metatype` and `is_constant` are used interchangeably :)

anvilfolk replied:
I think more specifically, if you use `resolve_datatype` or `reducer_identifier` on the `[A]` of `var my_array : Array [A]`, then the datatype will be a `metatype`, but you want elements to _not_ be `metatype`. Does that make sense at all?

vonagam replied:
Seems like it does not apply here then. If you have `[A, B]` where `A` and `B` are classes (not an instances) then it should be typed as an array of classes and not array of instances of those classes.

anvilfolk replied:
I'll trust you on it! Don't have capacity to dig very deep :)

My main concern was 
```GDScript

var arr : Array[A] = [A.new(), A.new()]
var my_var := arr[0]
```

I just want to make sure that the type of `my_var` is not a metatype. If it isn't, then we're all good!

vonagam replied:
Ah, yes, disregard previous comment. By that point if there is a specified datatype it will be called `resolve_datatype` on it. And [those lines](https://github.com/godotengine/godot/blob/6318320c04f348eee1df7a5dae3bad540ef1fdcc/modules/gdscript/gdscript_analyzer.cpp#L492-L494) make sure that the element type in not metatype.

anvilfolk replied:
Perfect! :D

anvilfolk commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -1929,49 +1885,50 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
+			continue;
+		}
+		if (is_type_compatible(this_element_type, element_type, false)) {
+			element_type = this_element_type;
```
Why is this needed? You're always doing a reverse check anyway.

Also, it seems simpler to do:
```
if (!is_type_compatible(element_type, this_element_type, false) && !is_type_compatible(this_element_type, element_type, false)) {
  return;
}
```

vonagam replied:
If you have array `[B, A]` then the inferred type is `Array[A]` even though `A` cannot be assigned to `B` (assuming that `B` extends `A`). [Here are those lines](https://github.com/godotengine/godot/blob/6318320c04f348eee1df7a5dae3bad540ef1fdcc/modules/gdscript/gdscript_analyzer.cpp#L1955-L1956) in original code.

anvilfolk replied:
Ah, I see. So this effectively goes up the class hierarchy as far up as it can? That makes sense to me then, and the code as-is looks good :)

vonagam replied:
Yes, but actually there is a small problem - it is not as smart, so if you have array `[B, C]` where `B` and `C` are separate descendants of `A`, then it will fail, will not infer that it is `Array[A]` unfortunately.

anvilfolk replied:
Sounds like a fun PR for another day though! :)

anvilfolk commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2866,10 +2823,10 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 				case GDScriptParser::ClassNode::Member::CONSTANT:
 					// For out-of-order resolution:
-					reduce_expression(member.constant->initializer);
+					resolve_constant(member.constant);
```
Love ittttt!!!

anvilfolk commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2912,8 +2869,8 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 							// For out-of-order resolution:
-							reduce_expression(member.constant->initializer);
-							p_identifier->set_datatype(member.get_datatype());
+							resolve_constant(member.constant);
```
yessssssssssssssssssssss

vonagam replied:
Thanks) 

I do think that other places like `resolve_variable` (with it's code copy in `resolve_class_interface`) and `resolve_parameter` can be unified with `resolve_constant` in the future under the same function too.

anvilfolk replied:
Yes, there's a lot of refactoring possible here, and I for one am excited for it :)

anvilfolk commented:
This generally looks good! I like it, and it seems to fix a bunch of things!

I think the things I am concerned about, which @kleonc mentioned (namely that arrays should be referenced, not copied), might be concerns for another PR.

What I will say though, is that I would love to see a handful of error cases added to the unit tests, to show that the typechecking works properly and catches errors - and to make sure there are no more regressions in the future! :)

vonagam commented:
True about failure tests. I would have preferred if there was something of opposite of `assert` like `fail` in tests, which expects an error and if there is none - logs the message. Awkward to split positive and negative checks into different files, breaks the code flow and logic.

As for referenced/copied, I think it is an important matter, needs to be agreed on before proceeding. Have a concern about `Array[String]` being equal to `Array[StringName]` like recent PR wanted, otherwise should not be a problem to change the behavior.

Cretezy commented:
Hi! Thanks for this PR, I will test locally.

I'd like to add a potential new case to fix (I don't know if this PR fixes this yet)

```gdscript
func test(items: Array[int]) -> void: pass

func _ready():
	# Works
	test([1])

	# Doesn't work
	test([1].map(func(x): x))
```

vonagam commented:
This is the same as #58636, last example of [this comment](https://github.com/godotengine/godot/issues/58636#issuecomment-1292030012). `map` returns untyped array and there is no conversion at the moment when you pass arrays as arguments.

Cretezy commented:
@vonagam Great, thanks. Happy to know it's tracked!

dalexeev commented:
Does this fix the following bug/issue?

```gdscript
extends Control

var _buttons: Array[Button]

func _ready() -> void:
    # Error: Trying to assign a typed array with an array of different type.
    _buttons = [$Button1, $Button2, $Button3]
```

While in some other cases this works:

```gdscript
extends Control

var _ints: Array[int]

func _ready() -> void:
    _ints = [1, 2, 3]
```

This is probably because the type is guessed, although it is not explicitly specified.

```gdscript
print([1, 2, 3].get_typed_builtin()) # 0 (TYPE_NIL)

var a := [1, 2, 3]
print(a.get_typed_builtin()) # 2 (TYPE_INT) !!!

var b: Array = [1, 2, 3]
print(b.get_typed_builtin()) # 0 (TYPE_NIL)

# Should be the opposite, default is `Array` and `Array[int]` should be explicit.
```

I think we are missing typed array literals (like `Array[int](1, 2, 3)`) to explicitly specify the type without an extra variable. Implicit guessing is confusing and inconsistent.

vonagam commented:
Sorry for delay in reply, @dalexeev. Added a fix for the shown issue of yours. 

Part of it was fixed before, but typed_assign in Array was still missing a case when you assign from an array of base classes (Node) to an array of extending classes (Button). (It is the same as the one where you assign variants to objects.)

Now it works. Also marked such assignments where a source array cannot be used without conversions/checks as type unsafe.

adamscott commented on core/variant/array.cpp:
```diff
@@ -202,53 +202,77 @@ uint32_t Array::recursive_hash(int recursion_count) const {
+	} else if (source_typed.type == Variant::OBJECT && typed.can_reference(source_typed)) {
+		// from extending classes to base classes
+		for (int i = 0; i < size; i++) {
+			data[i] = source[i];
```
Is this normal that this code is the same as the precedent block?
Could we join them?

vonagam replied:
Joined them.

adamscott reviewed (approved):
Seems good to me. Only true usage will tell, but the PR is really well made and the tests pass.

vnen commented:
It seems to me this still needs discussion on what the proper behavior should be. Only then we can say whether the implementation is optimal.

There's also a change in Array that I'm not really convinced. Before it used COW (copy-on-write) which only made a copy if the array changed, but now it's always making a copy. This probably has a performance impact. It's also one thing to be discussed: _when_ (or even _if_) an array should be copied instead of passed by reference.

vonagam commented:
I think that arrays should not be copied on usual assignment. As was said, arrays are reference types and when you receive an array then you can call `push_back` expecting original one to get element too. In my opinion the only question is convenient explicit api/syntax for converting arrays to other types (by copy).

```gdscript
var subclasses: Array[B] = []
func for_classes(classes: Array[A]): pass
```

What way for a user should there be to pass `subclasses` to `for_classes` (as copy) without too much of a hassle?

As for copy-on-write - that's sounds like a good idea. But I don't see any trace of it in master (I would not have removed such feature), am I missing something?

vonagam commented:
Another major problem with typed arrays is that they are not so much typed, meaning that all of their methods do return/accept same old variants. Array will not be the last parameterized thing in godot, so there is need to think about how to represent type parameters in info.

vonagam commented:
PR and its description are updated. Complete change of how things work.
It does not include `Array[int]()`. Can be added later.
Added implicit conversion for array literals.
Copy-on-write is back and public method `typed_assign` has a reasonable purpose now.
Will add tests once confirmed that described behavior is ok with everybody.

anvilfolk commented:
Super exciting!!!! All of it checks out with what we discussed in the meeting, I think!

The only thing that I'd like to discuss a bit more is this one

`var infer := [1, 2, 3] # type is Array, not Array[int]`

Wouldn't it make sense to try to infer the type of the array elements here as well, and if they're not all of the exact same type, default to `Variant` array?

dalexeev commented:
> The only thing that I'd like to discuss a bit more is this one
> 
> `var infer := [1, 2, 3] # type is Array, not Array[int]`

My opinion is that it should just be `Array` (`Array[Variant]`) since GDScript is a dynamic language by default and `Variant` should be used everywhere unless explicitly stated otherwise. `:=` asks us to guess the type, we guess it's an `Array`, but nothing more.

Otherwise, we will have problems such as:

```gdscript
print([1, 2, 3].get_typed_builtin()) # 0 (TYPE_NIL)

var a := [1, 2, 3]
print(a.get_typed_builtin()) # 2 (TYPE_INT) !!!

var b: Array = [1, 2, 3]
print(b.get_typed_builtin()) # 0 (TYPE_NIL)

# Should be the opposite, default is `Array` and `Array[int]` should be explicit.
```

and

```gdscript
var a = [1, 2, 3]
print(a.get_typed_builtin()) # 0 (TYPE_NIL)

var b := [1, 2, 3]
print(b.get_typed_builtin()) # 2 (TYPE_INT) !!!
```

vonagam commented:
Few things about it:
- By inferring typed array you remove ability to infer basic one (unless it is empty or complete mix). 
- `Array[A]` and  `Array[B]` are not compatible like before, so it important to get type right.
- Classes may have multiple base classes, it is not exact which one to infer, inferring most specific is an arbitrary decision.
- It complicates codebase.

I do think that decision on whenever or not to make array typed should be made by a user since there will be consequences. Most people should be ok with a basic array. Have no problem with putting that logic back if that's the consensus though.

anvilfolk commented:
I am convinced :)

vonagam commented:
What I am more interested in and what we have not discussed is the implicit conversion of an array literal to typed one.

It can be perceived as somewhat confusing:
```gdscript
var good_int: Array[int] = [0] # this works

var basic = [0]
var bad_int: Array[int] = basic # this does not
```

The reasoning is that when a literal converted we are sure that there are no other references to that array (so nobody can write into it) so it is ok to convert it to typed one.

While some possible confusion is a minus, there is a big plus that you don't have to write `Array[MyClassName]()` anytime you want to create new array for assignment into a variable, passing as an argument or returning from a function. So it will make them much less frictionless to use.

akien-mga commented:
> The reasoning is that when a literal converted we are sure that there are no other references to that array (so nobody can write into it) so it is ok to convert it to typed one.

Maybe the rule could be that implicit conversion only works with a const value? I.e. these two would work:
```gdscript
var good_int: Array[int] = [0] # this works

const CONSTANT = [0]
var good_int2: Array[int] = CONSTANT # this could work

var basic = [0]
var bad_int: Array[int] = basic # this does not
```
That being said it's a fairly theoretical use case, I don't see much value in it. So the current behavior which you outlined is probably sufficient.

vonagam commented:
Last one is already possible.

vonagam commented:
Added tests. Ready for review, I guess.

vonagam commented:
Improved a little the error message for when a wrong type of array is passed to an argument of typed array.

vonagam commented:
Improved a little the error message for when a wrong type of array is passed to an argument of typed array.

vonagam commented:
Updated new `update_const_expression_builtin_type` function to be more correct. The idea is to use it to convert constant expression values where it is possible to avoid doing it during runtime. In initializers, assignments, passing to arguments and returns.

```gdscript
var x: float = 0
x = 1 # can be converted

func print_float(y: float): print(y)
print_float(2) # can be converted

func make_float() -> float: return 3 # can be converted
print(make_float())
```

#70873 can use it for initializers, fixing bugs with constants in the process. I can apply it to those other places in different PR.

vonagam commented:
Fixed hashing for constant arrays - before type information were not included, so one could get array of wrong type from a constant if there was an array with same contents. For me it was empty arrays.

vonagam commented:
Ok, I think it is ready for review. @vnen take a look when you have a time.

Things that I know are still broken:

`resize` does not fill new data with variants to match the type. Is there a simple function to initialize default variant for a `Variant::Type`?

Custom classes cannot be used as an element type in constant arrays, so `const empty[A] = []` will not have right type, only builtin and native type will be set. (Fortunately besides creating empty arrays there is not much you can do in constant context with custom classes.)

I think `typed_assign` can be renamed to something simpler, just `assign` maybe.

vonagam commented:
Ok, fixed resizing. It did lead to crashes before, saw issue about it somewhere.

Renamed PR, at this point there are too many things that this PR fixes about arrays.

The only important issue remaining (that I know of) is about `script_type` not being properly available in const context or when class is in the same file. First one is connected to cache and nested classes. Second because compiler does not set `script_type` for the class in the same file as it is a reference and there was attempt to avoid self referencing the script.

vonagam commented:
Ok, added some more fixes, updated the description. Mainly about editor and writing/reading typed arrays into/from text.

Need to figure out how to represent custom user class in text (for writing/reading) and in property hint (for editor).

akien-mga commented:
This seems to change how default values are stringified, which breaks godot-cpp:
```
In file included from /home/runner/work/godot/godot/godot-cpp/gen/include/godot_cpp/classes/rd_vertex_attribute.hpp:38,
                 from /home/runner/work/godot/godot/godot-cpp/gen/src/classes/rd_vertex_attribute.cpp:33:
/home/runner/work/godot/godot/godot-cpp/gen/include/godot_cpp/classes/rendering_device.hpp: In lambda function:
Error: /home/runner/work/godot/godot/godot-cpp/gen/include/godot_cpp/classes/rendering_device.hpp:678:624: error: expected '{' before ',' token
  678 |  RID render_pipeline_create(const RID &shader, int64_t framebuffer_format, int64_t vertex_format, RenderingDevice::RenderPrimitive primitive, const Ref<RDPipelineRasterizationState> &rasterization_state, const Ref<RDPipelineMultisampleState> &multisample_state, const Ref<RDPipelineDepthStencilState> &stencil_state, const Ref<RDPipelineColorBlendState> &color_blend_state, BitField<RenderingDevice::PipelineDynamicStateFlags> dynamic_state_flags = (BitField<RenderingDevice::PipelineDynamicStateFlags>)0, uint32_t for_render_pass = 0, const TypedArray<RDPipelineSpecializationConstant> &specialization_constants = Array([], 24, &"RDPipelineSpecializationConstant", null));
      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                ^
```

And the docs:
```diff
diff --git a/doc/classes/SystemFont.xml b/doc/classes/SystemFont.xml
index 20bfd0d..d851176 100644
--- a/doc/classes/SystemFont.xml
+++ b/doc/classes/SystemFont.xml
@@ -19,7 +19,7 @@
 		<member name="antialiasing" type="int" setter="set_antialiasing" getter="get_antialiasing" enum="TextServer.FontAntialiasing" default="1">
 			Font anti-aliasing mode.
 		</member>
-		<member name="fallbacks" type="Font[]" setter="set_fallbacks" getter="get_fallbacks" default="[]">
+		<member name="fallbacks" type="Font[]" setter="set_fallbacks" getter="get_fallbacks" default="Array([], 24, &amp;&quot;Font&quot;, null)">
 			Array of fallback [Font]s.
 		</member>
 		<member name="font_italic" type="bool" setter="set_font_italic" getter="get_font_italic" default="false">
```

vonagam commented:
Yeah, I saw the error about usage in documentation. Need to think has to resolve it.

vnen commented:
I finally took some time to review this. Overall seems good to me. The only thing the is off is assignment of untyped -> typed arrays.

In particular, this case gives an error when it should not:
```gdscript
var arr: Array[int];
var i = 0
arr = [i]
```

It says `Cannot have untyped elements in a typed array` in the last line. This is problematic especially for people not using types (GDScript is still dynamically typed by default after all) when interacting with the engine methods or typed third-party addons. Here's an example of potential untyped code giving error, which is a bit contrived since not many things in Godot take typed arrays, but should be enough to illustrate the point:
```gdscript
extends CodeEdit
enum Tokens {
	TK_PERIOD,
}
func _ready(): 
	var token_map = {
		Tokens.TK_PERIOD: ".",
	}
	code_completion_prefixes = [token_map[Tokens.TK_PERIOD]]
```

While changing the dictionary from `var` to `const` works for this, it might not always be an option.

This is more odd considering untyped arrays and typed variant arrays are accepted. In the following image only the last line gives an error. The 3rd line in the function is even marked green:
```gdscript
func add_prefix(x):
	code_completion_prefixes += x
	var arr: Array[String] = code_completion_prefixes + [x]
	code_completion_prefixes = code_completion_prefixes + [x]
	code_completion_prefixes += [x]
```

Essentially, while my previous example gave a parser error, this does not:
```gdscript
var arr: Array[int];
var i = 0
var temp_arr = [i]
arr = temp_arr
```

When this will fail at runtime, but the first should not since it should (try to) create a typed array at runtime. This one even gives all green lines (and also gives error at runtime since the concatenation creates an untyped Array):

```gdscript
var arr: Array[int];
var i = 0
arr = [] + [i]
```

----

In summary, it's giving errors when it shouldn't and there are some cases where green lines are shown when they should not (in particular `typed_array = variant_array`).

I'll also have to discuss with other people the changes in core, especially in VariantParser (and Writer). I think we need to come up with a format that's both good for the documentation and for the serializing (or consider if it's worth splitting those for this case alone). 

For instance, using a number to store the Variant type in the scene file isn't great because it will make files unreadable if the things in `Variant::Type` enum changes order (unlikely to happen until maybe Godot 5, but even then we want to easily port scenes from one version to the next). It should probably write the type as text and have a map to get it back to the actual type, which also helps with documentation.

There's also the case of using `std::optional`. We've been a bit stringent in allowing `std` stuff in core for many reasons. I will have to consult others to see if this is okay.

I can't really comment for the changes in the inspector, I'm not familiar with that code.

reduz commented on core/variant/array.cpp:
```diff
@@ -123,6 +123,10 @@ bool Array::recursive_equal(const Array &p_array, int recursion_count) const {
 	if (_p == p_array._p) {
 		return true;
 	}
+	if (_p->typed != p_array._p->typed) {
```
Why would you not want this? I am not sure why you can't compare a typed array to untyped if the contents are the same.

reduz replied:
Remember that GDScript is by default untyped

vonagam replied:
It is an error to compare `PackedInt32Array` and `PackedInt64Array` by value.

And there is an inconsistency about nested elements in array:
```gdscript
print('a' == &'a') # true
print(['a'] == [&'a']) # false
```

So it is an arbitrary choice, I guess. Have no problem removing this.

vnen replied:
I believe arrays of different types can be considered different, even if empty. But untyped array should be comparable with any typed or untyped array by its contents.

```gdscript
var typed: Array[int] = [1, 2, 3]
var untyped = [1, 2, 3]
print(typed == untyped) # true
```

vonagam replied:
Related question: should type info be included in hash? So should `Array[StringName]()` and `Array[float]()` have the same hash or different one?

vonagam replied:
Ok, I'll just remove my additions to `recursive_equal` and `recursive_hash`.

vonagam commented:
> In particular, this case gives an error when it should not

Ok. Removed `Cannot have untyped elements in a typed array`, allowed variants inside literals. It will try to convert now (the same as calling constructor) and error in runtime if fails.

> The 3rd line in the function is even marked green

Ok. Fixed green lines for assignments (though there should be other places like returns and args and so on, can be fixed later).

> using a number to store the Variant type

I see `Variant::get_type_name` but I don't see an inverse function to get a type from a name.

> We've been a bit stringent in allowing `std` stuff in core for many reasons.

Just to be clear for others - `std::optional` usage here is limited to GDScript module, is not introduced into core.

h0lley commented:
presumably #71756 is another one for the list of what this will fix?

(note that enum typed arrays DID work up until beta 11 though).

adamscott commented:
I think we should merge this PR ASAP, then fix the issues that comes up afterwards. We're imminently close to release release candidates and we need to make sure that this fix is battle-tested.

vnen commented:
> > We've been a bit stringent in allowing `std` stuff in core for many reasons.
> 
> Just to be clear for others - `std::optional` usage here is limited to analyzer in GDScript module, is not introduced into core.

It seems there is still resistance among the maintainers (I too have some reservations because of future implications) so I'll ask you to remove the usage, instead use a `bool` return and pass the Variant by reference, or vice-versa. Being in the `core` folder or not does not make much of a difference since everything is in the main codebase and is included by default. 

> Unfortunately I do see `Variant::get_type_name` but don't see an inverse function to get a type from a name.

Yeah, that would need another map from name to type. @reduz suggested looping through the types and matching the name and if performance is an issue cache this statically accessible via a `Variant::get_type_from_name()` method.

There were suggestions of storing the value as if it were a typed constructor (e.g. `Array[int]([1, 2, 3])`) which does look good both on the scene and the docs. In case of custom scripts, they could be saved as `Array[ExtResource("1_w6xmu")]([])`. The VariantParser should be able to detect the `ExtResource(` or `SubResource(`  prefix and treat it accordingly. This is looks to me like the best option.

vonagam commented:
But I don't think that `Array[int]([1, 2, 3])` will work in generated C++.

vnen commented:
> But I don't think that `Array[int]([1, 2, 3])` will work in generated C++.

The bindings generator in godot-cpp can adapt this syntax to generate what is appropriate for C++. The current format in the PR (`Array([1, 2, 3], 2, &"", null)`) is also not valid C++ and would also need an update in the bindings generator.

vonagam commented:
Is it up to me fix then? What representation should be used for `default_value` in gdextension dump?
`godot-cpp` in another repository and I think [`correct_default_value`](https://github.com/godotengine/godot-cpp/blob/ae1afba8d126e2b577ba358f17b2843787f71ef1/binding_generator.py#L2118) would need to be updated.

akien-mga commented:
@bruvzg or @vnen might be able to help fix the godot-cpp bindings generator.
The correct C++ syntax to replace e.g. `Array[RID]([])` would be `TypedArray<RID>()`. There's probably no need to support custom script classes there as they wouldn't be used in the generated API.

vonagam commented:
True, since it is `TypedArray<T>` it is clearly C++ specific and no changes to gdextension dump is needed.

For docs I kept current format. For that `DocData::get_default_value_string` was added. Since docs do have their own notation for typed arrays anyway and to keep things cleaner.

Besides C++ changes needed in `godot-cpp` repository this one is ready. One small thing though:

I would prefer to rename `typed_assign`. There is no point to keep `typed` in the name now, it is a method in the same category as `push_back`, `append` and `append_array` (not in category of `set_typed`, `is_typed` or `get_typed_builtin`). It does more conversions than those, but the same behavior can be applied to them later.

akien-mga commented:
Now that https://github.com/godotengine/godot-cpp/pull/1018 is merged, this may pass CI after a rebase.

vonagam commented:
I've rebased the other two PRs. After they are merged will rebase this one and resolve an easy conflict with "constant conversions".

As for rename, I though about which name would be suitable, simply `assign` is ok, but when added later to `Dictionary` maybe too close in name to `merge`. Looked around and I think `copy` is a good one, in C# this is a description for [`Array.Copy`](https://learn.microsoft.com/en-us/dotnet/api/system.array.copy?view=net-7.0):
```
Copies a range of elements in one Array to another Array and performs type casting and boxing as required.
```

Which matches intended behavior of the method.

vonagam commented:
I rebased and renamed the method. godot-cpp still has a problem because it has [`include/godot_cpp/variant/typed_array.hpp`](https://github.com/godotengine/godot-cpp/blob/0f3a0913f100dcd988b538582431d064f46c4e17/include/godot_cpp/variant/typed_array.hpp) which this PR changes, the old one used `typed_assign` in two places which were replace by `_ref` call.

h0lley commented:
I haven't been able to reconstruct it in a minimal project, but loading a larger project with this PR, the data serialized in many `.tres` files cannot be loaded (is just being ignored - no error messages). when then re-saving those tres files, the data is lost.

for instance, a `SpriteFrames` resource plugged into a `AnimatedSprite2D` node is just dropped.

in another case, one of my resources serializes sub resources in a typed array: `@export var arr: Array[MyRes]`, which are all being dropped; the inspector displays an empty array and when saving the resource, the data is lost.

vonagam commented:
When types are incompatible during setting up an instance from a scene they are silently ignored. That's how it is, yes. So all typed arrays currently serialized in scene files as untyped arrays will be dropped.

One option to soften this is to allow temporary conversion from untyped array to typed one and then later in separate PR disallow such conversion. But it will not guarantee that all untyped arrays were properly resaved and some people skip releases.

And to be clear, after you work with your project and save new arrays (with new format) they are no longer being dropped, only old ones, right?

h0lley commented:
yes once serialized with the new syntax it's all good.
unfortunately there doesn't seem to be an easy way to identify and fix affected files with a simple search & replace.

real world example from previous to new:

```
conversation = [SubResource("2"), SubResource("3")]
```

```
conversation = Array[Resource("res://generic/resources/dialogue/dialogue_item.gd")]([SubResource("Resource_fvm2a"), SubResource("Resource_qswwh")])
```

akien-mga commented:
```
conversation = Array[Resource("res://generic/resources/dialogue/dialogue_item.gd")]([SubResource("Resource_fvm2a"), SubResource("Resource_qswwh")])
```

Shouldn't `Resource("res://generic/resources/dialogue/dialogue_item.gd")` also be a `SubResource()` or `ExtResource()` instead of `Resource(path)`?

I haven't tested on a big project myself, but I think we do need to provide an upgrade path for existing scenes. Maybe VariantParser can keep support for the old syntax as additional compatibility code, while VariantWriter would write the new?

vonagam commented:
> VariantParser can keep support for the old syntax as additional compatibility code

Old syntax is supported and as before produces untyped arrays. `VariantParser` cannot produce typed arrays from `[]` strings, since it doesn't know what type of an array is expected.

Currently in people's scene files all arrays are saved as untyped ones. With this PR there will be no implicit conversion from untyped one to typed one on assignment as before. So incompatible.

vnen commented:
> As for rename, I though about which name would be suitable, simply `assign` is ok, but when added later to `Dictionary` maybe too close in name to `merge`. Looked around and I think `copy` is a good one, in C# this is a description for [`Array.Copy`](https://learn.microsoft.com/en-us/dotnet/api/system.array.copy?view=net-7.0):
> 
> ```
> Copies a range of elements in one Array to another Array and performs type casting and boxing as required.
> ```
> 
> Which matches intended behavior of the method.

I don't like the name `copy()` because it sounds to me like it returns a copy of the array (same as `duplicate()`). I think `assign()` makes more sense since you are pretty much performing an assignment operation. Not everyone is familiar with C# to understand the `copy` name right away.

vonagam commented:
`copy` and `clone` are generic names for such methods. `clone` does what `duplicate` does in gdscript. C# is more relevant than some other languages because it is a second officially supported language. For array `assign` might work, but for dictionaries (or other types) it might become confusing - `merge` and `assign` (where there is `Object.assign` in js which what `merge` does). `copy` has a required argument so it will be clear that it is not what `duplicate` does.

kleonc commented:
Note that C#'s `Array.Copy` is a static method which takes in both source and destination arrays (and it's capable of copying only a specific range of elements, possibly starting at specific destination index). A non-static version is [`Array.CopyTo`](https://learn.microsoft.com/en-us/dotnet/api/system.array.copyto?view=net-7.0#system-array-copyto(system-array-system-int32)) which copies the whole array to the specified destination (array + start index). But what's important: these methods actually do perform copying.

In case of "our method" it might end up just assigning the internal array which is copy-on-write and hence actual copying can happen later. So `copy` (or `copy_from` which I think would be a better name) would be a little misleading. On the other hand `assign` does make sense to me, that's pretty much what this method does.

vonagam commented:
> In case of "our method" it might end up just assigning the internal array which is copy-on-write and hence actual copying can happen later. So copy (or copy_from which I think would be a better name) would be a little misleading.

I don't how it is misleading. It changes contents to match those of source array. That's it. copy-on-write is just an internal optimization/implementation, has nothing to do with behavior.

I'm okay with changing name to `assign`, I asked yesterday about it, nobody was against, for some reason now it is a problem.

vonagam commented:
I added such code to `SceneState::instantiate` in `PackedScene`:
```cpp
if (value.get_type() == Variant::ARRAY) {
  Array set_array = value;
  bool is_get_valid = false;
  Variant get_value = node->get(snames[nprops[j].name], &is_get_valid);
  if (is_get_valid && get_value.get_type() == Variant::ARRAY) {
    Array get_array = get_value;
    if (!set_array.is_same_typed(get_array)) {
      value = Array(set_array, get_array.get_typed_builtin(), get_array.get_typed_class_name(), get_array.get_typed_script());
    }
  }
}
```
So if an array being set but it is possible to get a default one and the element types are different then try to convert the incoming array to type of currently present one.

Not super ideal - proper is to use `get_property_list` and parse the hint to get the type (to handle setter-only exports, but much more code though), not only packed scenes use parser/writer.

vonagam commented:
Added the same logic to `ResourceLoaderText::load()`.

akien-mga commented:
Thanks!

anvilfolk commented:
Congratulations on this huge PR! I know it was a loooooooooot of work with a looooooooot of subtleties, but so very important! Very very excited this got in! :) :) :)
